pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+func ParseRemoteReadRequestWithoutCosumingBody(req *http.Request) (url.Values, error) {
+	params := make(url.Values)
+
+	if req.Body == nil {
+		return params, nil
+	}
+
+	bodyBytes, err := util.ReadRequestBodyWithoutConsuming(req)
+
+	if req.Header.Get("Content-Encoding") == "snappy" {
+		bodyBytes, err = snappy.Decode(nil, bodyBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Parse the request data.
+	remoteReadRequest := prompb.ReadRequest{}
+	if err := proto.Unmarshal(bodyBytes, &remoteReadRequest); err != nil {
+		return nil, err
+	}
+
+	queries := remoteReadRequest.GetQueries()
+
+	for i, query := range queries {
+		params.Add(fmt.Sprintf("%d_start", i), fmt.Sprintf("%d", query.GetStartTimestampMs()))
+		params.Add(fmt.Sprintf("%d_end", i), fmt.Sprintf("%d", query.GetEndTimestampMs()))
+
+		matchersStrings := make([]string, 0, len(query.Matchers))
+		matchers, err := remote.FromLabelMatchers(query.Matchers)
+		if err != nil {
+			return nil, err
+		}
+		for _, m := range matchers {
+			matchersStrings = append(matchersStrings, m.String())
+		}
+		params.Add(fmt.Sprintf("%d_matchers", i), strings.Join(matchersStrings, ","))
+
+		// TODO handle hints.
```

Comment:
Is this something we're going to handle before merging? What's the blocker?

Reply:
Need to figure out what the fields mean and decide if we want to keep them or just dump all?
```
    StepMs               int64    `protobuf:"varint,1,opt,name=step_ms,json=stepMs,proto3" json:"step_ms,omitempty"`
    Func                 string   `protobuf:"bytes,2,opt,name=func,proto3" json:"func,omitempty"`
    StartMs              int64    `protobuf:"varint,3,opt,name=start_ms,json=startMs,proto3" json:"start_ms,omitempty"`
    EndMs                int64    `protobuf:"varint,4,opt,name=end_ms,json=endMs,proto3" json:"end_ms,omitempty"`
    Grouping             []string `protobuf:"bytes,5,rep,name=grouping,proto3" json:"grouping,omitempty"`
    By                   bool     `protobuf:"varint,6,opt,name=by,proto3" json:"by,omitempty"`
    RangeMs              int64    `protobuf:"varint,7,opt,name=range_ms,json=rangeMs,proto3" json:"range_ms,omitempty"`
```

Reply:
How about just json-marshaling them (as they have `omitempty`) and setting that as `hints=` as the first approach?

Reply:
I think Oleg idea is simple yet effective for the logging purpose.

Reply:
I saw this after I implemented json marshal :)